packages/typescript-checker/src/compilers/compiler-with-watch.ts

```diff
@@ -0,0 +1,177 @@
+    guardTSVersion();
+    this.guardTSConfigFileExists();
+    const buildModeEnabled = determineBuildModeEnabled(this.tsconfigFile);
+
+    const host = ts.createSolutionBuilderWithWatchHost(
+      {
+        ...ts.sys,
+        readFile: (fileName) => {
+          const content = this.fs.getFile(fileName)?.content;
+          if (content && this.allTSConfigFiles.has(path.resolve(fileName))) {
+            return this.adjustTSConfigFile(fileName, content, buildModeEnabled);
+          }
+          return content;
+        },
+        getModifiedTime: (pathName: string) => {
+          return this.fs.getFile(pathName)?.modifiedTime;
+        },
+        watchFile: (filePath: string, callback: ts.FileWatcherCallback) => {
+          const file = this.fs.getFile(filePath);
+
+          if (file) file.watcher = callback;
+
+          return {
+            close: () => {
+              delete this.fs.getFile(filePath)!.watcher;
+            },
+          };
+        },
+        writeFile: (fileName, data) => {
+          this.fs.writeFile(fileName, data);
+        },
+        watchDirectory: (): ts.FileWatcher => {
+          // this is used to see if new files are added to a directory. Can safely be ignored for mutation testing.
+          return {
+            // eslint-disable-next-line @typescript-eslint/no-empty-function
+            close() {},
+          };
+        },
+      },
+      undefined,
+      (error) => {
+        this.currentErrors.push(error);
+      },
+      (status) => {
+        this.log.debug(status.messageText.toString());
+      },
+      () => {
+        this.currentTask.resolve();
+      }
+    );
+
+    const dependencyFiles: SourceFiles = {};
+
+    host.afterProgramEmitAndDiagnostics = (program: ts.EmitAndSemanticDiagnosticsBuilderProgram) => {
+      program
+        .getSourceFiles()
+        .filter(filterDependency)
+        .forEach((file) => {
+          dependencyFiles[file.fileName] = {
+            fileName: toPosixFileName(file.fileName),
+            imports: new Set(
+              program
+                .getAllDependencies(file)
+                .filter((importFile) => !importFile.includes('/node_modules/') && file.fileName !== importFile)
+                .flatMap((importFile) => this.resolveFilename(importFile))
+            ),
+            importedBy: new Set(),
+          };
+        });
+    };
+
+    const compiler = ts.createSolutionBuilderWithWatch(host, [this.tsconfigFile], {});
+
+    compiler.build();
+
+    const errors = await this.check();
+
+    // Function should not be called after the first check is done
+    host.afterProgramEmitAndDiagnostics = undefined;
+
+    for (const file in dependencyFiles) {
+      const imports = dependencyFiles[file].imports;
+
+      for (const importFile of imports) {
+        dependencyFiles[importFile]?.importedBy.add(file);
+      }
+    }
+
+    function filterDependency(file: ts.SourceFile) {
```

Comment:
Can we move this to a `helper` or something? Also might need a new name. What are we filtering?